### PR TITLE
throw when report steps has start time later than end time

### PR DESCRIPTION
**Pull request by GitPaean**

### CODE REVIEW

**Comment by GitPaean** [comment]
@bska , you probably have some comments and suggestions about this. I just created the PR here to begin the discussion.

**Comment by bska** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
`mpi.msim` test failure is a little more involved to figure out. 

But basically it was a problematically setup restart scenario and the PR revealed the mistake.

**Comment by GitPaean** [comment]
For the test failure `mpi.msim`, Without figuring out the exact cause yet, it looks like the deck schedule is a combination of many `TSTEP`s and with one `DATES` somewhere in the middle. When restarting from the 50th report step, it might add the whole schedule from the 50th report step, the time steps of `TSTEP` is added on top of it, then we have a `DATES` in the middle causing the problem. 

This part of the code is not trival, it is great someone more familiar with the RST facility can give some suggestions. @bska, any comment?

File: src/opm/input/eclipse/Schedule/Schedule.cpp [hunk 1]
```diff
@@ -1261,6 +1261,12 @@ File {} line {}.)", pattern, location.keyword, location.filename, location.linen
     }
 
     double Schedule::stepLength(std::size_t timeStep) const {
+        if (this->snapshots[timeStep].end_time() < this->snapshots[timeStep].start_time()) {
+            const auto& start_time = Schedule::formatDate(std::chrono::system_clock::to_time_t(this->snapshots[timeStep].start_time()));
+            const auto& end_time = Schedule::formatDate(std::chrono::system_clock::to_time_t(this->snapshots[timeStep].end_time()));
+            throw std::runtime_error(fmt::format(" Report step {0} has the start time {1} later than the end time {2},\n"
+                                                 "can be due to the wrong way doing restarting (forgetting SKIPREST?) or bad setup in the SCHEDULE section.", timeStep, start_time, end_time));
+        }
```
Comment (bska) [inline src/opm/input/eclipse/Schedule/Schedule.cpp:1269]: I think I'd probably write this as
```C++
const auto start = this->snapshots[timeStep].start_time();
const auto end = this->snapshots[timeStep].end_time();

if (start > end) {
    throw std::invalid_argument {
        fmt::format(" Report step {} has start time after end time,\n"
                    "   * Start time = {:%d-%b-%Y %H:%M:%S}\n"
                    "   * End time   = {:%d-%b-%Y %H:%M:%S}.\n"
                    " Possibly due to inconsistent RESTART/SKIPREST settings.",
                    timeStep + 1,
                    fmt::gmtime(TimeService::to_time_t(start)),
                    fmt::gmtime(TimeService::to_time_t(end)))
    }
}

return std::chrono::duration_cast<std::chrono::seconds>(end - start).count();
```
Note that `fmt::gmtime()` and the conversion specifiers for `std::tm` require that we
```C++
#include <fmt/chrono.h>
```

**Review by bska — COMMENTED** [review]
We have a mechanism in place to detect a missing `DATES` record in the context of a `RESTART`ed simulation run using `SKIPREST` (see PR #2959), so this PR complements the consistency checking by detecting the `SKIPREST` keyword being absent altogether.

While I would personally implement the diagnostic slightly differently, especially using the [`TimeService`](https://github.com/OPM/opm-common/blob/d27888009159e7ac8d296e855c41f9acf01411ff/opm/common/utility/TimeService.hpp#L34-L49) facilities, I do appreciate you taking the time to get into this part of the code.

**Comment by GitPaean** [comment]
Thanks @bska for the suggestions.  This PR targets the situation that restarting with deck schedule `DATES` without specifying `SKIPREST`, 

At the same time,  the master branch can not detect the following two situations, 
1. TSTEP with negative values
2. DATES, while with messed up order of dates. 

For the first one, just modifying the first TSTEP with a negative sign to be `-31`, we can the following output
```
Report step  0/120 at day 0/3588, date = 01-Jan-2015

Report step  1/120 at day -31/3588, date = 01-Dec-2014

Starting time step 0, stepsize 1 days, at day -31/-3, date = 01-Dec-2014
Program threw an exception: Elapsed time (-30.0 [DAYS]) must not precede previous elapsed time (0.0 [DAYS]). Incorrect restart time?
```

For the second one with `DATES`, we get the following output by modifying the `DATES` in BASE_MODEL.DATA, we get the following output, 
```
Report step  0/12 at day 0/364, date = 01-Jan-2000

Report step  1/12 at day -334/364, date = 01-Feb-1999

Starting time step 0, stepsize 1 days, at day -334/60, date = 01-Feb-1999
Program threw an exception: Elapsed time (-333.0 [DAYS]) must not precede previous elapsed time (0.0 [DAYS]). Incorrect restart time?
```

Should we detect them in a much earlier stage, for example parsing stage?

**Comment by GitPaean** [comment]
jenkins build this please

File: src/opm/input/eclipse/Schedule/Schedule.cpp [hunk 2]
```diff
@@ -1276,8 +1277,19 @@ File {} line {}.)", pattern, location.keyword, location.filename, location.linen
...
+        if (start > end) {
+            throw std::invalid_argument {
+                    fmt::format(" Report step {} has start time after end time,\n"
+                                "   * Start time = {:%d-%b-%Y %H:%M:%S}\n"
+                                "   * End time   = {:%d-%b-%Y %H:%M:%S}.\n"
+                                " Possibly due to inconsistent RESTART/SKIPREST settings.",
+                                timeStep + 1,
+                                fmt::gmtime(TimeService::to_time_t(start)),
+                                fmt::gmtime(TimeService::to_time_t(end))) };
+        };
```
Comment (bska) [inline src/opm/input/eclipse/Schedule/Schedule.cpp:1291]: Tiny nit, but a trailing `;` on the closing brace of `if()`, while not an error, does look a little weird here.

File: src/opm/input/eclipse/Schedule/ScheduleDeck.cpp [hunk 3]
```diff
@@ -332,6 +342,13 @@ void ScheduleDeck::add_block(ScheduleTimeType time_type, const time_point& t, Sc
 void ScheduleDeck::add_TSTEP(const DeckKeyword& TSTEPKeyword, ScheduleDeckContext& context) {
     const auto &item = TSTEPKeyword.getRecord(0).getItem(0);
     for (size_t itemIndex = 0; itemIndex < item.data_size(); itemIndex++) {
+        {
+            const auto tstep = item.get<double>(itemIndex);
+            if (tstep < 0) {
+                const auto msg = fmt::format("a negatvie TSTEP value {} is input", tstep);
```
Comment (bska) [inline src/opm/input/eclipse/Schedule/ScheduleDeck.cpp:348]: Misprint: "negatVIe" versus "negatIVe".

**Review by bska — APPROVED** [review]
Looks good for the most part except for two tiny nits.  Once you fix those I'll be happy to merge this into the master branch.

**Comment by GitPaean** [comment]
jenkins build this please

File: src/opm/input/eclipse/Schedule/ScheduleDeck.cpp (hunk 3 above)
Comment (GitPaean) [reply src/opm/input/eclipse/Schedule/ScheduleDeck.cpp]: Thanks for the good catch! Both comments are addressed.

**Comment by GitPaean** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Thanks for the updates.  This looks good to me now and the build check is green.  I'll merge into master.

**Pull request merged**